src/Host/UserProfile/Impl/RUserProfileService.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Security.Principal;
+using System.ServiceProcess;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.R.Host.UserProfile {
+    partial class RUserProfileService : ServiceBase {
+        public RUserProfileService() {
+            InitializeComponent();
+        }
+
+        private ManualResetEvent _workerdone;
+        private CancellationTokenSource _cts;
+        protected override void OnStart(string[] args) {
+            base.OnStart(args);
+            _cts = new CancellationTokenSource();
+            _workerdone = new ManualResetEvent(false);
+            Task.WhenAny(CreateProfileWorkerAsync(_cts.Token));
+        }
+
+        protected override void OnStop() {
+            base.OnStop();
+            _cts.Cancel();
+            _workerdone.WaitOne(TimeSpan.FromSeconds(10));
+        }
+
+        async Task CreateProfileWorkerAsync(CancellationToken ct) {
```

Comment:
Same thing about async as on the other end of the pipe. I think it's worse here, actually, because it's more likely that broker will crash (it's more complex), it doesn't have the monitor process, and it doesn't own the server end of the pipe (so you won't notice that it's gone). So if broker crashes between writes for username/domain/password, or before it could read the result, this will get stuck in the middle of the loop, and won't respond to service stop requests.